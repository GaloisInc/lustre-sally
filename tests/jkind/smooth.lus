-- This example illustrates the benefits of the -smooth flag

node delay(signal : bool) returns (out : bool);
var
  counter : int;
let
  counter = if signal then 1 + (0 -> pre counter) else 0;
  out = counter <= 10;
tel;

node main(x, y, z, w : int) returns (cex : bool);
let
  cex = delay(x < y and y < z and z < w and x < w);
  --%PROPERTY cex;
tel;
