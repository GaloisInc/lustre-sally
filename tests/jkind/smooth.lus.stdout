Core Lustre
===========

node (x : int, y : int, z : int, w : int)
  returns (cex)
  assumes ()
  shows (cex)
let
  signal : bool
    = And(And(And(Lt(x, y), Lt(y, z)), Lt(z, w)), Lt(x, w))
   
  pre_counter : int
    = pre counter
   
  init_pre_counter : int
    = 0 -> pre_counter
   
  counter : int
    = ITE(signal, Add(1, init_pre_counter), 0)
   
  cex : bool
    = Leq(counter, 10)
tel
Sally Model
===========

(define-state-type
   S
   (
     (cex Bool)
     (counter Int)
     (signal Bool)
     (w Int)
     (x Int)
     (y Int)
     (z Int)
     (|gal-initializing| Bool)
     (|init_pre_counter:init| Bool)
     (|init_pre_counter| Int)
     (|pre_counter| Int))
   (
     (w Int)
     (x Int)
     (y Int)
     (z Int)))
(define-transition-system
   TS
   S
   (and
      (= |gal-initializing| true)
      (= |init_pre_counter:init| true))
   (and
      (= next.|gal-initializing| false)
      (= next.x input.x)
      (= next.y input.y)
      (= next.z input.z)
      (= next.w input.w)
      (=
         next.signal
         (and
            (< next.x next.y)
            (< next.y next.z)
            (< next.z next.w)
            (< next.x next.w)))
      (= next.|pre_counter| state.counter)
      (ite
         state.|init_pre_counter:init|
         (= next.|init_pre_counter| 0)
         (= next.|init_pre_counter| next.|pre_counter|))
      (= next.|init_pre_counter:init| false)
      (=
         next.counter
         (ite
            next.signal
            (+ 1 next.|init_pre_counter|)
            0))
      (=
         next.cex
         (<= next.counter 10))))

(query
   TS
   (or |gal-initializing| cex))

[Lustre] Validating properties:
[Lustre] Property cex... considering simultaneous states to depth 0
considering simultaneous states to depth 0
considering simultaneous states to depth 1
considering simultaneous states to depth 1
considering simultaneous states to depth 2
considering simultaneous states to depth 2
considering simultaneous states to depth 3
considering simultaneous states to depth 3
considering simultaneous states to depth 4
considering simultaneous states to depth 4
considering simultaneous states to depth 5
considering simultaneous states to depth 5
considering simultaneous states to depth 6
considering simultaneous states to depth 6
considering simultaneous states to depth 7
considering simultaneous states to depth 7
considering simultaneous states to depth 8
considering simultaneous states to depth 8
considering simultaneous states to depth 9
considering simultaneous states to depth 9
considering simultaneous states to depth 10
considering simultaneous states to depth 10
counter-example search depth 0
counter-example search depth 1
counter-example search depth 2
counter-example search depth 3
counter-example search depth 4
counter-example search depth 5
counter-example search depth 6
counter-example search depth 7
counter-example search depth 8
counter-example search depth 9
counter-example search depth 10
[[1;93mUnknown[0m] Valid up to depth 10
[Lustre] Model status: [[1;93mUnknown[0m] 
